File: CRM/src/main/kotlin/it/polito/wa2/g07/crm/entities/lab03/JobOffer.kt
package it.polito.wa2.g07.crm.entities.lab03


import it.polito.wa2.g07.crm.entities.lab03.Customer
import jakarta.persistence.ElementCollection
import jakarta.persistence.Entity
import jakarta.persistence.Id
import jakarta.persistence.ManyToOne

import jakarta.persistence.*

import com.fasterxml.jackson.annotation.JsonIgnore

import kotlin.time.Duration
import kotlin.time.DurationUnit

enum class OfferStatus {
    CREATED {
        override val compatibleStatuses: Collection<OfferStatus>
            get() = setOf(SELECTION_PHASE, ABORTED)
    },
    SELECTION_PHASE {
        override val compatibleStatuses: Collection<OfferStatus>
            get() = setOf(CANDIDATE_PROPOSAL, ABORTED)
    },
    CANDIDATE_PROPOSAL {
        override val compatibleStatuses: Collection<OfferStatus>
            get() = setOf(CONSOLIDATED, SELECTION_PHASE, ABORTED)
    },
    CONSOLIDATED {
        override val compatibleStatuses: Collection<OfferStatus>
            get() = setOf(DONE, SELECTION_PHASE, ABORTED)
    },
    DONE {
        override val compatibleStatuses: Collection<OfferStatus>
            get() = setOf(SELECTION_PHASE)
    },
    ABORTED {
        override val compatibleStatuses: Collection<OfferStatus>
            get() = setOf()
    };

    protected abstract val compatibleStatuses: Collection<OfferStatus>
    fun canUpdateTo(status: OfferStatus) = this.compatibleStatuses.contains(status)
}

@Entity
class JobOffer(


    @ElementCollection(fetch = FetchType.EAGER)
    var requiredSkills: MutableSet<String> = mutableSetOf(),

    var duration: Long,

    var description: String,

    var status: OfferStatus = OfferStatus.CREATED,

    var notes: String? = null,
) {
    @ManyToOne
    lateinit var customer: Customer

    companion object {
        const val PROFIT_MARGIN = 0.2
    }

    @Id
    @GeneratedValue
    var offerId: Long = 0L

    @ManyToOne(cascade = [CascadeType.ALL])
    var professional: Professional? = null
        set(professional) {
            professional?.jobOffers?.add(this)
            field = professional
        }

    val value: Double?
        get() = when(this.professional){
            null -> null
            //  null -> throw IllegalStateException("Value cannot be calculated if the job offer has no professional!")
            else -> this.duration * this.professional!!.dailyRate * PROFIT_MARGIN
        }


}

File: Frontend/src/types/JobOffer.ts
import { Pageable } from "./Pageable.ts";
import { ReducedProfessional } from "./professional.ts";
import { ReducedCustomer } from "./customer.ts";

export type JobOfferStatus =
  | "CREATED"
  | "SELECTION_PHASE"
  | "CANDIDATE_PROPOSAL"
  | "CONSOLIDATED"
  | "DONE"
  | "ABORTED";

export interface ReducedJobOffer {
  id: number;
  description: string;
  offerStatus: JobOfferStatus;
  customer: ReducedCustomer | null;
  professional: ReducedProfessional | null;
}

export interface JobOfferResponse {
  content: ReducedJobOffer[];
  pageable: Pageable;
  totalPages: number;
}

export interface JobOffer {
  id: number;
  description: string;
  customer: ReducedCustomer;
  requiredSkills: string[];
  duration: number;
  notes?: string;
  professional: ReducedProfessional;
  value: number;
  offerStatus: JobOfferStatus;
  candidates: ReducedProfessional[];
  refusedCandidates: ReducedProfessional[];
}

export interface JobOfferCreate {
  description: string;
  requiredSkills: string[];
  duration: number;
  notes: string | null | undefined;
}

export interface JobOfferUpdateStatus {
  status: JobOfferStatus;
  professionalId?: number; //ReducedProfessional;
}
